### PR TITLE
 Improve Memory Efficiency for SparseEncoder models 

### PR DESCRIPTION
**Pull request by arthurbr11**
Hello!

## Pull Request overview
* This pull request refactors the `SparseEncoder`  to simplify functionality and improve efficiency. Key changes include removing the `output_value` parameter from the `encode` method, optimizing the `forward` method in `MLMTransformer`, and introducing memory-efficient pooling in `SpladePooling`.

### Refactoring and Simplification:
* Removed the `output_value` parameter from the `encode` method in `SparseEncoder`, simplifying the logic by always returning sentence embeddings. This eliminates handling token embeddings as it's memory consuming and shouldn't be used here. 

* Optimized the `forward` method in `MLMTransformer` by directly assigning the MLM logits to `features["token_embeddings"]`, removing intermediate variables and redundant comments.

### Efficiency Improvements:
* Added a `memory_efficient` option to the `forward` method in `SpladePooling`, enabling chunk-wise computation for large inputs. This reduces memory usage while maintaining functionality for max and sum pooling strategies. 

* Implemented a chunk-based pooling mechanism in `SpladePooling`, ensuring efficient computation of pooled scores (either max or sum) while supporting activation functions like ReLU and log1p_relu.

## TODO
* Fix the nondeterministic behavior of the training and them make sure they give same output in both configurations 
* Benchmark speed/memory usage in both config with small to large models with small to large vocab_size
 

- Arthur BRESNU

### CODE REVIEW

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 1]
```diff
@@ -1022,6 +999,34 @@ def max_seq_length(self, value) -> None:
...
+        """
+        Sets the chunk size of the SpladePooling module, if present.
+        """
+        found_splade = False
+        for mod in self._modules.values():
+            if isinstance(mod, SpladePooling):
+                mod.chunk_size = value
+                found_splade = True
+                break
+        if not found_splade:
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:1027]:
```suggestion
                break
        else:
```
The for-else is one of the least known about Python features, but it works great for exactly this. The `else` branch executes only if the `for` branch ended normally, i.e. without a `break`.

File: sentence_transformers/sparse_encoder/SparseEncoder.py [hunk 2]
```diff
@@ -1022,6 +999,34 @@ def max_seq_length(self, value) -> None:
         """
         self._first_module().max_seq_length = value
 
+    @property
+    def chunk_size(self) -> int | None:
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/SparseEncoder.py:1003]: Maybe it should be the `splade_pooling_chunk_size` or something? And it's not very clear what this does right now

File: sentence_transformers/sparse_encoder/models/MLMTransformer.py [hunk 3]
```diff
@@ -106,13 +106,9 @@ def forward(self, features: dict[str, torch.Tensor]) -> dict[str, torch.Tensor]:
...
         }
-        # Get MLM outputs
-        mlm_outputs = self.auto_model(**trans_features)
-
-        # Get the MLM head logits (shape: batch_size, seq_length, vocab_size)
-        mlm_logits = mlm_outputs.logits
-
-        features["token_embeddings"] = mlm_logits
+        features["token_embeddings"] = self.auto_model(
+            **trans_features, output_hidden_states=False, output_attentions=False
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/MLMTransformer.py:110]: I don't think we should specify the `output_hidden_states` and `output_attentions`, these are False by default unless the user specifies otherwise in the `transformers` `AutoConfig`, and then I don't think we need to manually specify it ourselves.

File: sentence_transformers/sparse_encoder/models/SpladePooling.py [hunk 4]
```diff
@@ -28,14 +32,20 @@ class SpladePooling(Module):
                 - `relu`: ReLU activation (standard in all Splade models).
                 - `log1p_relu`: log(1 + ReLU(x)) variant used in Opensearch Splade models see arxiv.org/pdf/2504.14839.
         word_embedding_dimension (int, optional): Dimensionality of the output embeddings (if needed).
+        chunk_size (int, optional): Size of chunks when processing tokens. If None, processes entire sequence at once.
+            Using smaller chunks reduces memory usage but may impact speed performance. Default is None.
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/SpladePooling.py:36]: `chunks` is not very clear, normally I would think that this chunks in the batch dimension and not in the sequence length dimension.

File: sentence_transformers/sparse_encoder/models/SpladePooling.py [hunk 5]
```diff
@@ -28,14 +32,20 @@ class SpladePooling(Module):
...
     """
 
     SPLADE_POOLING_MODES = ("sum", "max")
     SPLADE_ACTIVATION = ["relu", "log1p_relu"]
     config_keys: list[str] = ["pooling_strategy", "activation_function", "word_embedding_dimension"]
 
     def __init__(
-        self, pooling_strategy: str = "max", activation_function="relu", word_embedding_dimension: int = None
+        self,
+        pooling_strategy: str = "max",
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/SpladePooling.py:45]:
```suggestion
        pooling_strategy: Literal["max", "sum"] = "max",
```
This might be nice actually?

File: sentence_transformers/sparse_encoder/models/SpladePooling.py [hunk 6]
```diff
@@ -28,14 +32,20 @@ class SpladePooling(Module):
...
+        activation_function="relu",
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/SpladePooling.py:46]:
```suggestion
        activation_function: Literal["relu", "log1p_relu"] = "relu",
```
idem? What do you think?

File: sentence_transformers/sparse_encoder/models/SpladePooling.py [hunk 7]
```diff
@@ -45,32 +55,77 @@ def __init__(
...
-        else:  # sum
-            pooled_scores = torch.sum(splade_scores, dim=1)  # shape: batch_size, vocab_size
+            pooled_scores = torch.full((batch_size, vocab_s), float("-inf"), dtype=mlm_logits.dtype, device=device)
+        elif self.pooling_strategy == "sum":
+            pooled_scores = torch.zeros((batch_size, vocab_s), dtype=mlm_logits.dtype, device=device)
+        else:
+            raise ValueError(f"Unsupported pooling_strategy: {self.pooling_strategy}")
+
+        # Process in chunks if chunk_size is set, otherwise process the entire sequence at once
+        chunk_size = seq_len if self.chunk_size is None else self.chunk_size
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/SpladePooling.py:91]:
```suggestion
        chunk_size = seq_len if (self.chunk_size is None or self.chunk_size <= 0) else self.chunk_size
```
I think this is safest to avoid an issue with the `range`

File: sentence_transformers/sparse_encoder/models/SpladePooling.py [hunk 8]
```diff
@@ -45,32 +55,77 @@ def __init__(
...
+                if self.activation_function == "relu":
+                    if not self.training:
+                        current_chunk_transformed = masked_current_chunk_logits.relu_().log1p_()
+                    else:
+                        current_chunk_transformed = torch.log1p(torch.relu(masked_current_chunk_logits))
+                elif self.activation_function == "log1p_relu":
+                    if not self.training:
+                        current_chunk_transformed = masked_current_chunk_logits.relu_().log1p_().log1p_()
+                    else:
+                        current_chunk_transformed = torch.log1p(torch.log1p(torch.relu(masked_current_chunk_logits)))
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/SpladePooling.py:109]:
```suggestion
                if not self.training:
                    current_chunk_transformed = masked_current_chunk_logits.relu_().log1p_()
                else:
                    current_chunk_transformed = masked_current_chunk_logits.relu().log1p()
                # With "log1p_relu", we apply a second log1p
                if self.activation_function == "log1p_relu":
                    if not self.training:
                        current_chunk_transformed = current_chunk_transformed.log1p_()
                    else:
                        current_chunk_transformed = current_chunk_transformed.log1p()
```
What do you think of this simplification?

File: sentence_transformers/sparse_encoder/models/SpladePooling.py [hunk 9]
```diff
@@ -45,32 +55,77 @@ def __init__(
...
+
+                current_chunk_transformed = current_chunk_transformed.to(mlm_logits.dtype)
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/SpladePooling.py:111]: I believe there's nothing that changed the dtype until now, so this should still always already be `mlm_logits.dtype`, right? So then we might not need this

File: sentence_transformers/sparse_encoder/models/SpladePooling.py [hunk 10]
```diff
@@ -45,32 +55,77 @@ def __init__(
...
+
+                if self.pooling_strategy == "max":
+                    chunk_pooled = torch.max(current_chunk_transformed, dim=1)[0]
+                    pooled_scores = torch.maximum(pooled_scores, chunk_pooled)
+                else:
+                    chunk_pooled = torch.sum(current_chunk_transformed, dim=1)
+                    pooled_scores += chunk_pooled
+            except RuntimeError as e:
```
Comment (tomaarsen) [inline sentence_transformers/sparse_encoder/models/SpladePooling.py:119]: Nice! Have you been able to test this?

File: sentence_transformers/sparse_encoder/SparseEncoder.py (hunk 1 above)
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/SparseEncoder.py:1027]: Didn't know it at all, thanks !

File: sentence_transformers/sparse_encoder/models/SpladePooling.py (hunk 6 above)
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/models/SpladePooling.py:46]: Yes i like it, it will make it easier to change

File: sentence_transformers/sparse_encoder/models/SpladePooling.py (hunk 8 above)
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/models/SpladePooling.py:109]: It is indeed simplifying the code, so i like it !

File: sentence_transformers/sparse_encoder/models/SpladePooling.py (hunk 10 above)
Comment (arthurbr11) [reply sentence_transformers/sparse_encoder/models/SpladePooling.py:119]: Yes i did worked well !

**Comment by tomaarsen** [comment]
Update: I figured out that you can't train with `log1p_` (as we know), but we *can* train with `relu_`, so we can use e.g.:

```python
                current_chunk_transformed = masked_current_chunk_logits.relu_()
                if not self.training:
                    current_chunk_transformed = current_chunk_transformed.log1p_()
                else:
                    current_chunk_transformed = current_chunk_transformed.log1p()
                # With "log1p_relu", we apply a second log1p
                if self.activation_function == "log1p_relu":
                    if not self.training:
                        current_chunk_transformed = current_chunk_transformed.log1p_()
                    else:
                        current_chunk_transformed = current_chunk_transformed.log1p()
```

and perhaps we decrease memory usage when training?

- Tom Aarsen
